haskell/private/cc_wrapper.py.tpl

```diff
@@ -0,0 +1,812 @@
+
+    """
+    LINK = "link"
+    COMPILE = "compile"
+    PRINT_FILE_NAME = "print-file-name"
+
+    def __init__(self, args):
+        """Parse the given arguments into an Args object.
+
+        - Shortens library search paths.
+        - Detects the requested action.
+        - Keeps rpath arguments for further processing when linking.
+        - Keeps print-file-name arguments for further processing.
+
+        Args:
+          args: Iterable over command-line arguments.
+
+        """
+        self.action = Args.LINK
+        self.print_file_name = None
+        self.libraries = []
+        self.library_paths = []
+        self.rpaths = []
+        self.output = None
+        self._prev_ld_arg = None
+
+        self.args = list(self._handle_args(args))
+
+        if not self.linking:
+            # We don't expect rpath arguments if not linking, however, just in
+            # case, forward them if we don't mean to modify them.
+            self.args.extend(rpath_args(self.rpaths))
+
+    @property
+    def linking(self):
+        """Whether this is a link invocation."""
+        return self.action == Args.LINK and self.output is not None
+
+    @property
+    def compiling(self):
+        """Whether this is a compile invocation."""
+        return self.action == Args.COMPILE
+
+    @property
+    def printing_file_name(self):
+        """Whether this is a print-file-name invocation."""
+        return self.action == Args.PRINT_FILE_NAME and self.print_file_name is not None
+
+    def _handle_args(self, args):
+        """Argument handling pipeline.
+
+        Args:
+          args: Iterable, command-line arguments.
+
+        Yields:
+          Transformed command-line arguments.
+
+        """
+        args = iter(args)
+        for arg in args:
+            out = []
+            # Poor man's pattern matching: Each handler function takes the
+            # current argument, the stream of up-coming arguments, and a
+            # reference to the list of arguments to forward. The handler must
+            # return True if it consumes the argument, and return False if
+            # another handler should consume the argument.
+            if self._handle_output(arg, args, out):
+                pass
+            elif self._handle_library(arg, args, out):
+                pass
+            elif self._handle_library_path(arg, args, out):
+                pass
+            elif self._handle_linker_arg(arg, args, out):
+                pass
+            elif self._handle_print_file_name(arg, args, out):
+                pass
+            elif self._handle_compile(arg, args, out):
+                pass
+            else:
+                yield arg
+
+            for out_arg in out:
+                yield out_arg
+
+    def _handle_output(self, arg, args, out):
+        if arg == "-o":
+            # Remember the output filename.
+            self.output = next(args)
+            out.extend(["-o", self.output])
```

Comment:
I’d go for a more structured approach here and add `"-o"`, `"-l"` and such not here, but in the final “pretty-printing phase” of converting `Args` to a GHC command line.

That is, just save the data (`self.output` in this match) and add the lower-level structure of the args later.

Reply:
The less structured approach is intentional. This wrapper tries hard to maintain the relative order of the command-line arguments, structuring all flags and only then pretty-printing them could change that order. The only exception are arguments that we intend to modify and where we need to have seen all command-line flags before we can modify them, namely `--print-file-name` and `-rpath`.

Generally, it is not safe to change the order of command-line arguments. E.g. with static library linking the relative order of the arguments matters. Consider `-l foo libbar.a -l baz`, where `libfoo` and `libbaz` are also static libraries. If we would structure and then pretty print all arguments, then we'd end up with something like `-l foo -l baz libbar.a`, which changes meaning.

We could structure and also remember the position of each argument and then puzzle them back together in the right order. But, I don't see that the benefit outweighs the additional complexity.

Reply:
True